src/vision/status_quo/alan_hates_writing_a_stream.md

```diff
@@ -0,0 +1,194 @@
+
+Eventually (how?), Alan realizes that the types he's depending on are `Unpin`, and so he doesn't need to worry about the unsafe stuff. It's just a more-annoying pointer type.
+
+[self-ref]: https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+
+#### State Machine
+
+With `Pin` hopefully ignored, Alan next notices that in the imperative style he wanted originally, he didn't need to explicitly keep track of state. The state was simply the imperative order of the function. But in a `poll` function, the state isn't saved by the compiler. Alan finds blog posts about the dark ages of Futures 0.1, when even `Future` was written with a "state machine".
+
+He thinks about his stream's states, and settles on the following structure:
+
+```rust
+struct SigningFile {
+    state: State,
+    file: ReaderStream,
+    sig: Signature,
+}
+
+enum State {
+    File,
+    Sign,
+}
+```
+
+Now he tries to write the `poll_next` method, checking readiness of individual steps and proceeding to the next state, while grumbling away the weird `Pin` noise:
+
+```rust
+todo!("is this worth showing in the story, or just noise?")
+```
+
+It turns out it was more complicated than Alan thought (the author made this same mistake). The `digest` method of `Signature` is `async`, _and_ it consumes the signature, so the state machine needs to be adjusted. The signature needs to be able to be moved out, and it needs to be able to store a future from an `async fn`. Trying to figure out how to represent that in the type system was difficult (expand?), but Alan eventually learns to just store a `Pin<Box<dyn Future>>`, wondering if the `Pin` there is important.
+
+```rust
+struct SigningFile {
+    state: State,
+}
+
+enum State {
+    File(ReaderStream, Signature),
+    Signing(Pin<Box<dyn Future<Item = Vec<u8>>>>),
+    Done,
+}
+```
+
+(show how gross this codes gets?)
+
+Oh well, at least it _works_, right?
+
+#### Wakers
+
+So far, Alan hasn't paid too much attention to `Context` and `Poll`. It's been fine to simply pass them along untouched. There's a confusing bug in his state machine. Let's look more closely:
+
+```rust
+// zooming in!
+match ready!(file.poll_next(cx)) {
+    Some(val) => {
+        me.sig.as_mut().unwrap().push(val.len()));
+        return Poll::Ready(Some(val));
+    },
+    None => {
+        me.state = State::Sign;
+        // oops!
+        return Poll::Pending;
+    }
+}
+```
+
+In one of the branches, the state is changed, and `Poll::Pending` is returned. Alan assumes that the task will be polled again with the new state. But, since the file was done (and has returned `Poll::Ready`), there was actually no waker registered to wake the task again. So his stream just hangs forever.
```

Comment:
The code uses the `ready!` macro. How did Alan land on using that? Did its use potentially obscure things even more?

Reply:
I've updated that Alan remembers `ready!` from the futures 0.1 blog posts. I don't believe it obscures things (but I could be blind by my own knowledge...)